From df34690242fd07e9db8c4bcd631c139ef1f61544 Mon Sep 17 00:00:00 2001
From: Alejandro Abdelnur <tucu@apache.org>
Date: Wed, 16 Jan 2013 01:07:35 +0000
Subject: [PATCH 004/596] MR1: MAPREDUCE-4924. flakey test:
 org.apache.hadoop.mapred.TestClusterMRNotification.testMR.
 (rkanter via tucu)

git-svn-id: https://svn.apache.org/repos/asf/hadoop/common/branches/branch-1@1433782 13f79535-47bb-0310-9956-ffa450edef68
(cherry picked from commit d0fb29450104fb814e2b492af5db8581f3a02109)
(cherry picked from commit 6b22cdf4b73f7b6f5f40da4e8b4f53c6a4b77ce0)
(cherry picked from commit 5ebee4cb29477ee3625edfae0837465ef35c2ade)
(cherry picked from commit 6a3a3c34a7517cd7eb25510f652d1a9017062898)
(cherry picked from commit 9116b4bc002ef9d2e55ea661a36fe4e87bd49ad3)
---
 .../apache/hadoop/mapred/NotificationTestCase.java |   23 +++++++++++---------
 1 file changed, 13 insertions(+), 10 deletions(-)

diff --git a/hadoop-mapreduce1-project/src/test/org/apache/hadoop/mapred/NotificationTestCase.java b/hadoop-mapreduce1-project/src/test/org/apache/hadoop/mapred/NotificationTestCase.java
index 9565c96..282a857 100644
--- a/hadoop-mapreduce1-project/src/test/org/apache/hadoop/mapred/NotificationTestCase.java
+++ b/hadoop-mapreduce1-project/src/test/org/apache/hadoop/mapred/NotificationTestCase.java
@@ -96,7 +96,7 @@ private void stopHttpServer() throws Exception {
   }
 
   public static class NotificationServlet extends HttpServlet {
-    public static int counter = 0;
+    public static volatile int counter = 0;
 
     protected void doGet(HttpServletRequest req, HttpServletResponse res)
       throws ServletException, IOException {
@@ -160,9 +160,10 @@ protected void tearDown() throws Exception {
   public void testMR() throws Exception {
     System.out.println(launchWordCount(this.createJobConf(),
                                        "a b c d e f g h", 1, 1));
-    synchronized(Thread.currentThread()) {
-      stdPrintln("Sleeping for 2 seconds to give time for retry");
-      Thread.currentThread().sleep(2000);
+    boolean keepTrying = true;
+    for (int tries = 0; tries < 30 && keepTrying; tries++) {
+      Thread.sleep(50);
+      keepTrying = !(NotificationServlet.counter == 2);
     }
     assertEquals(2, NotificationServlet.counter);
     
@@ -180,18 +181,20 @@ public void testMR() throws Exception {
     // run a job with KILLED status
     System.out.println(UtilsForTests.runJobKill(this.createJobConf(), inDir,
                                                 outDir).getID());
-    synchronized(Thread.currentThread()) {
-      stdPrintln("Sleeping for 2 seconds to give time for retry");
-      Thread.currentThread().sleep(2000);
+    keepTrying = true;
+    for (int tries = 0; tries < 30 && keepTrying; tries++) {
+      Thread.sleep(50);
+      keepTrying = !(NotificationServlet.counter == 4);
     }
     assertEquals(4, NotificationServlet.counter);
     
     // run a job with FAILED status
     System.out.println(UtilsForTests.runJobFail(this.createJobConf(), inDir,
                                                 outDir).getID());
-    synchronized(Thread.currentThread()) {
-      stdPrintln("Sleeping for 2 seconds to give time for retry");
-      Thread.currentThread().sleep(2000);
+    keepTrying = true;
+    for (int tries = 0; tries < 30 && keepTrying; tries++) {
+      Thread.sleep(50);
+      keepTrying = !(NotificationServlet.counter == 6);
     }
     assertEquals(6, NotificationServlet.counter);
   }
-- 
1.7.9.5

